### PR TITLE
session: Circumvent user shell stdout noise

### PR DESCRIPTION
**Pull request by martinpitt**
session: Circumvent user shell stdout noise
  
Commit a9375441df7 caused Cockpit logins to fail if the user's ~/.bashrc
or ~/.profile writes anything to stdout. That interferes with the
protocol and causes an error like

    incorrect protocol: received invalid length prefix

To avoid this, route the shell's stdout to stderr (so that we can still
see it in the logs) and pass session's stdout to cockpit-bridge through
fd 3. This assumes that the shell supports `>&3` input redirection,
which is the case for any POSIX shell (ash, bash, ksh, dash, zsh) and at
least tcsh.

https://bugzilla.redhat.com/show_bug.cgi?id=1710604

### CODE REVIEW

**Comment by martinpitt** [comment]
Note: This should work on both *-distropkg. The regression didn't yet reach 7.6 and 8.0, so these shouldn't care what the user's shell is or does.

**Comment by martinpitt** [comment]
Reworked to route shell stdout to stderr instead of /dev/null. This makes it appear in the logs and is easier to do.

**Comment by martinpitt** [comment]
Still broken on atomic, it seems docker doesn't get along with the stdout noise:

    dockerd-current[1218]: [Invalid UTF-8] noise-rc-out

The fd redirection works fine though. As that doesn't seem to be a problem with cockpit-session, but just between ~/.bashrc and docker, I'll skip the test on Atomic.

**Comment by martinpitt** [comment]
Yay, check-login succeeded everywhere now, so this is ready for review.

File: test/verify/check-login [hunk 1]
```diff
@@ -133,6 +133,16 @@ account    required     pam_succeed_if.so user ingroup %s""" % m.get_admin_group
             b.open("/system")
             b.wait_not_visible("#option-group")
 
+        if not m.atomic_image:
+            # login with user shell that prints some stdout/err noise
+            # having stdout output in ~/.bashrc confuses docker, so don't run on Atomic
+            m.execute("echo 'echo noise-rc-out; echo noise-rc-err >&2' >> ~admin/.bashrc")
+            m.execute("echo 'echo noise-profile-out; echo noise-profile-err >&2' >> ~admin/.profile")
+            login("admin", "foobar")
+            with b.wait_timeout(10):
```
Comment (mvollmer) [inline test/verify/check-login]: What's the reason to use a faster timeout here?  I think we shouldn't change it.

Comment (martinpitt) [reply test/verify/check-login]: No particular reason really, it's just copy&paste from all the other similar instances in this test. I'm happy to change it wholesale.

Comment (martinpitt) [reply test/verify/check-login]: Pushed!

Comment (mvollmer) [reply test/verify/check-login]: Ah, I see.  Thanks!

**Comment by martinpitt** [comment]
The NFS test failures are real, but unrelated, fixed in PR #11986 .

**Pull request merged**